### ISSUE TITLE
typechecker panic when loop variable converted implicitly to different types

### ISSUE THREAD

**Issue by charles-cooper** [vyperlang/vyper#2772]
### Version Information

* vyper Version (output of `vyper --version`): 228b5bd920e9d56a8cb1a3262d7d60554cde265d
* Python Version (output of `python --version`): 3.8.10

### What's your issue about?

minimal repro:
```vyper
MAX_DEPTH: constant(uint256) = 6

@external
@view
def _getNodeDepth(_seed : uint256) -> uint8:
    for i in range(9):
        depth_up : uint256 = pow_mod256(convert(MAX_DEPTH, uint256), i)  # `i` inferred to be uint256
        return i  # i inferred to be uint8
    return 0
```
raises the following
```
  File "/home/charles/vyper/vyper/builtin_functions/functions.py", line 174, in fetch_call_return
    validate_expected_type(node.args[0], ValueTypeDefinition())
  File "/home/charles/vyper/vyper/semantics/validation/utils.py", line 458, in validate_expected_type
    f"Expected {expected_str} but literal can only be cast as {given_str}", node
  File "/home/charles/vyper/vyper/semantics/types/bases.py", line 540, in __repr__
    return self._id
AttributeError: 'ValueTypeDefinition' object has no attribute '_id'
```

**Comment by charles-cooper** [comment]
should be closed when https://github.com/vyperlang/vyper/pull/2817 is merged

**Comment by charles-cooper** [comment]
actually, this seems to be fixed as of v0.3.3.
